chia/wallet/wallet_node.py

```diff
@@ -1129,18 +1130,15 @@ async def new_peak_wallet(self, new_peak: wallet_protocol.NewPeakWallet, peer: W
 
     async def new_peak_from_trusted(self, new_peak_hb: HeaderBlock, latest_timestamp: uint64, peer: WSChiaConnection):
         current_height: uint32 = await self.wallet_state_manager.blockchain.get_finished_sync_up_to()
-        async with self.wallet_state_manager.lock:
+        async with self.wallet_state_manager.set_sync_mode(new_peak_hb.height):
             await self.wallet_state_manager.blockchain.set_peak_block(new_peak_hb, latest_timestamp)
             # Disconnect from all untrusted peers if our local node is trusted and synced
             await self.disconnect_and_stop_wpeers()
             # Sync to trusted node if we haven't done so yet. As long as we have synced once (and not
             # disconnected), we assume that the full node will continue to give us state updates, so we do
             # not need to resync.
             if peer.peer_node_id not in self.synced_peers:
-                if new_peak_hb.height - current_height > self.LONG_SYNC_THRESHOLD:
```

Comment:
you ended up removing this logic. It doesn't look right. I would have expected the context manager to just wrap the call to `long_sync()`, just like we do today.

It looks like an important distinction to allow holding the wallet state manager mutex while also not setting `sync_mode`.

Reply:
Yeah i removed that intentionally because `LONG_SYNC_THRESHOLD` is 300 and we roll back 256 blocks anyway in this case so why not just set the sync flag.. because, thats what we do?

Reply:
I'm not following this logic. Under what circumstance do we roll back 256 blocks?
That still means we can be closer than 300 blocks to the peak, in which case this flag did not used to be set. My impression is that this is distinguishing between long and short sync. is that not the case?

Reply:
I think we should get a better understanding of what the intention was, and make less assumptions by just reading the code. Testing this patch and explore the different scenarios would help build confidence in the patch being correct. Someone clearly meant to not set this flag sometimes, for some reason. It seems odd, but it's risky to assume that it was the original author who was confused, and not us.

Reply:
It's not that i didn't try to understand the intention for this line. If you search for the usage of `sync_mode` you can see it's so far only used in wallet RPC in two places, one to show the sync status in `get_sync_status` and the other to show the cached/empty balances during sync in `get_wallet_balance`.

So basically when we are doing the long sync with the wallet state manager lock held but the sync mode not set we are blocking `get_wallet_balance` calls and with this the GUI until the sync is done.

You can just just restart your trusted wallet on main and with this PR to see the behaviour difference during the initial sync. The main points i figured are:

### On `main`

- The balance view in the GUI shows loading spinners all over the place during the initial sync
- The NFT page only shows a big loading spinner until the initial sync is done
- The connections view just shows a big loading spinner until the initial sync is done

### With this PR

- The balance view shows 0 during the initial sync
- The NFT page shows the existing NFTs during the initial sync with a "loading" indication in the corner
- The connections screen shows the connections properly during the initial sync

(updated the initial description with the above)

So basically the UI is blocked and shows loading spinners all over the place during syncing if we don't set the sync mode (like currently on `main`). In my case here this means a blocked UI on startup for at least 15s with my current wallet and with this PR it shows at least some of the available information right after start.

The 0 balances are probably improvable by another PR to maybe generate the cached balanced before the sync on startup but for now its at least the same behaviour like if we are >300 blocks behind on `main` so it should not lead to confusion or questions? I asked mariano about his intention and i report back if there was something important to it but i feel like it just makes sense to merge this either way?

Reply:
> The 0 balances are probably improvable by another PR to maybe generate the cached balanced before the sync on startup 

Like here: #14631. If you apply this patch on top of this PR the wallet shows the last balance on start during the catchup sync also instead of showing just the loading spinners (`main`) or 0s (this PR only).